source/common/upstream/cluster_manager_impl.h

```diff
@@ -405,13 +446,20 @@ class ClusterManagerImpl : public ClusterManager, Logger::Loggable<Logger::Id::u
 
     bool blockUpdate(uint64_t hash) { return !added_via_api_ || config_hash_ == hash; }
 
-    LoadBalancerFactorySharedPtr loadBalancerFactory() {
+    // ClusterManagerCluster
+    Cluster& cluster() override { return *cluster_; }
+    LoadBalancerFactorySharedPtr loadBalancerFactory() override {
       if (thread_aware_lb_ != nullptr) {
         return thread_aware_lb_->factory();
       } else {
         return nullptr;
       }
     }
+    bool needsAddOrUpdate() override {
+      const bool old_value = added_or_updated_;
+      added_or_updated_ = true;
+      return !old_value;
```

Comment:
This seems a little bit brittle? This seems to assume that the caller will ensure that the cluster will be added/updated after calling this once, which could change down the line

Reply:
I wasn't crazy about this either, and originally I didn't have it and tried to pass a bool to postThreadLocalUpdate, but the problem is there is implied state around whether a cluster has been added/updated or not which has to be stored somewhere. I will see if I can think of anything better to do here. Would you rather there be an explicit setter/getter?

Reply:
Setter/Getter seems a bit nicer to me, just to avoid what I would have expected to be const from the name to actually mutate internal state. Renaming this might also make it better to make it clearer that its not idempotent?